papis/importer.py

```diff
@@ -0,0 +1,108 @@
+import papis
+import papis.utils
+import logging
+import os.path
+from stevedore import extension
+
+logger = logging.getLogger('importer')
+
+
+class Context:
+    files = []
+    data = dict()
+
+
+class Importer:
+    """This is the base class for every importer"""
+
+    def __init__(self, uri="", name="", ctx=Context()):
+        """
+        :param uri: uri
+        :type  uri: str
+        :param name: Name of the importer
+        :type  name: str
+        """
+        assert(isinstance(uri, str))
+        assert(isinstance(name, str))
+        assert(isinstance(ctx, Context))
+        self.uri = uri
+        self.name = name or os.path.basename(__file__)
+        self.logger = logging.getLogger("importer:{0}".format(self.name))
+        self.ctx = ctx
+
+    @classmethod
+    def match(uri):
+        """This method should be called to know if a given uri matches
+        the importer or not.
+
+        For example, a valid match for archive would be:
+        .. code:: python
+
+            return re.match(r".*arxiv.org.*", uri)
+
+        it will return something that is true if it matches and something
+        falsely otherwise.
+
+        :param uri: uri where the document should be retrieved from.
+        :type  uri: str
+        """
+        raise NotImplementedError(
+            "Matching uri not implemented for this downloader"
+        )
+
+    def fetch(self):
+        """
+        can return a dict to update the document with
+        """
+        raise NotImplementedError()
+
+    def __str__(self):
+        return 'Importer({0}, uri={1})'.format(self.name, self.uri)
+
+
+def stevedore_error_handler(manager, entrypoint, exception):
+    logger.error("Error while loading entrypoint [%s]" % entrypoint)
+    logger.error(exception)
```

Comment:
I think we should move this to __init__ as it is used in almost every plugin-based code.

Reply:
Agreed, originally I wanted to do a `plugin.py` central package, however I messed up with stevedore, stevedore is sensitive about some things, also specially about initializing twice an extension manager, that's why everywhere we should write a function like _create_importer_mgr
that only creates the manager if has not been created previously.